### PR TITLE
LDAP KeyStore

### PR DESCRIPTION
**Pull request by emre-aydin**

### CODE REVIEW

File: src/ManagementCenter.md [hunk 1]
```diff
@@ -139,7 +139,8 @@ You can use your existing LDAP server for authentication/authorization on Manage
 Provide the details in this form for your LDAP server:
 
 - **URL:** URL of your LDAP server, including schema (`ldap://` or `ldaps://`) and port.
-- **Search base DN:** Base DN to use for searching users/groups.
+- **Distinguished name (DN) of user:** DN of a user that has admin privileges on the LDAP server. Will be used to connect to the server when authenticating users.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ...LDAP server. It is used to connect....

File: src/ManagementCenter.md [hunk 2]
```diff
@@ -139,7 +139,8 @@ You can use your existing LDAP server for authentication/authorization on Manage
 Provide the details in this form for your LDAP server:
 
 - **URL:** URL of your LDAP server, including schema (`ldap://` or `ldaps://`) and port.
-- **Search base DN:** Base DN to use for searching users/groups.
+- **Distinguished name (DN) of user:** DN of a user that has admin privileges on the LDAP server. Will be used to connect to the server when authenticating users. 
+- **Password:** Password for the same user.- **Search base DN:** Base DN to use for searching users/groups.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: - **Password:** Password for the same user.
- **Search base DN:** Base DN to be used for searching users/groups.

File: src/ManagementCenter.md [hunk 3]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
 
 Once configured, LDAP settings are saved in a file named `ldap.properties` under the `mancenter` folder mentioned in the previous section. If you want to update your settings afterwards, you need to update `ldap.properties` file and click "Reload Security Config" button on the login page.
  
+#### Password encryption
```
Comment (Serdaro) [inline src/ManagementCenter.md]: #### Password Encryption

File: src/ManagementCenter.md [hunk 4]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+ 
+By default, the password that you use in LDAP configuration is saved on the `ldap.properties` file in clear text. This might pose a security risk. To store the LDAP password in encrypted form, we offer the following two options:
+
+- **Provide a KeyStore password:** This will create and manage a Java KeyStore under the Management Center home directory. The LDAP password will be stored in this KeyStore in encrypted form.
+- **Configure an external Java KeyStore:** This will use an existing Java KeyStore. This option might also be used to store the password in an HSM that provides a Java KeyStore API.
+
+When you do either, the LDAP password you enter on the initial configuration UI dialog will be stored in encrypted form in a Java KeyStore instead of the `ldap.properties` file.
+
+##### Providing a master key for encryption
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ##### Providing a Master Key for Encryption

File: src/ManagementCenter.md [hunk 5]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+
+There are two ways to provide a master key for encryption:
+
+- If you deploy Management Center on an application server, you need to set `MC_KEYSTORE_PASS` environment variable before starting Management Center. This option is less secure. You should clear the environment variable once you make sure you can log in with your LDAP credentials to minimize the security risk.
+- If you're starting Management Center from the commandline, you can start it with `-Dhazelcast.mc.askKeyStorePassword`. Management Center will ask for the KeyStore password upon start and use it as a password for the KeyStore it creates. This option is more secure as it only stores the KeyStore password in the memory.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ....from the command line, you can start it.....

File: src/ManagementCenter.md [hunk 6]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+
+By default, Management Center will create a Java KeyStore file under the Management Center home directory with the name `mancenter.jceks`. You can change location of this file by using the `-Dhazelcast.mc.keyStore.path=/path/to/keyStore.jceks` JVM argument.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ....You can change the location of this.....

File: src/ManagementCenter.md [hunk 7]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+ 
+##### Configuring an external Java KeyStore
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ##### Configuring an External Java KeyStore

File: src/ManagementCenter.md [hunk 8]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+
+If you don't want Management Center to create a KeyStore for you and use an existing one that you've created before (or an HSM), set the following JVM arguments when starting Management Center:
+
+* `-Dhazelcast.mc.useExistingKeyStore=true`: Enables use of an existing KeyStore.
+* `-Dhazelcast.mc.existingKeyStore.path=/path/to/existing/keyStore.jceks`: Path to the KeyStore. Might be empty if using an HSM.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ..to the KeyStore. It can be empty if you use an HSM.

File: src/ManagementCenter.md [hunk 9]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+* `-Dhazelcast.mc.existingKeyStore.pass=somepass`: Password for the KeyStore. Might be empty if HSM provides another means of unlocking HSM.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ...KeyStore. You do not have to set it if HSM provides another means to unlock HSM.

File: src/ManagementCenter.md [hunk 10]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+* `-Dhazelcast.mc.existingKeyStore.type=JCEKS`: Type of the KeyStore.
+* `-Dhazelcast.mc.existingKeyStore.provider=com.yourprovider.MyProvider`: Provider of the KeyStore. Leave empty to use the system provider. Specify the class name of your HSM's `java.security.Provider` implementation if you're using an HSM.
```
Comment (Serdaro) [inline src/ManagementCenter.md]: ...implementation if you use an HSM.

File: src/ManagementCenter.md [hunk 11]
```diff
@@ -151,6 +152,40 @@ Provide the details in this form for your LDAP server:
...
+
+* `-Dhazelcast.mc.useExistingKeyStore=true`: Enables use of an existing KeyStore.
+* `-Dhazelcast.mc.existingKeyStore.path=/path/to/existing/keyStore.jceks`: Path to the KeyStore. Might be empty if using an HSM.
+* `-Dhazelcast.mc.existingKeyStore.pass=somepass`: Password for the KeyStore. Might be empty if HSM provides another means of unlocking HSM.
+* `-Dhazelcast.mc.existingKeyStore.type=JCEKS`: Type of the KeyStore.
+* `-Dhazelcast.mc.existingKeyStore.provider=com.yourprovider.MyProvider`: Provider of the KeyStore. Leave empty to use the system provider. Specify the class name of your HSM's `java.security.Provider` implementation if you're using an HSM. 
+
+![image](images/NoteSmall.jpg) ***NOTE:*** *Make sure your KeyStore supports storing `SecretKey`s. *
+
+#### Updating encrypted password
```
Comment (Serdaro) [inline src/ManagementCenter.md]: #### Updating Encrypted Passwords

**Comment by emre-aydin** [comment]
@Serdaro Thanks for the review. Addressed all items on the review and updated the PR.

**Comment by Serdaro** [comment]
@emre-aydin , Is this update only for maintenance?

**Pull request merged**